p2p/nodestate/nodestate.go

```diff
@@ -0,0 +1,966 @@
+func (a Flags) Or(b Flags) Flags { return flagOp(a, b, true, true, true) }
+
+// Xor returns the set of flags present in either a or b but not both
+func (a Flags) Xor(b Flags) Flags { return flagOp(a, b, true, true, false) }
+
+// HasAll returns true if b is a subset of a
+func (a Flags) HasAll(b Flags) bool { return flagOp(a, b, false, true, false).mask == 0 }
+
+// HasNone returns true if a and b have no shared flags
+func (a Flags) HasNone(b Flags) bool { return flagOp(a, b, false, false, true).mask == 0 }
+
+// Equals returns true if a and b have the same flags set
+func (a Flags) Equals(b Flags) bool { return flagOp(a, b, true, true, false).mask == 0 }
+
+// IsEmpty returns true if a has no flags set
+func (a Flags) IsEmpty() bool { return a.mask == 0 }
+
+// MergeFlags merges multiple sets of state flags
+func MergeFlags(list ...Flags) Flags {
+	if len(list) == 0 {
+		return Flags{}
+	}
+	res := list[0]
+	for i := 1; i < len(list); i++ {
+		res = res.Or(list[i])
+	}
+	return res
+}
+
+// String returns a list of the names of the flags specified in the bit mask
+func (f Flags) String() string {
+	if f.mask == 0 {
+		return "[]"
+	}
+	s := "["
+	comma := false
+	for index, flag := range f.setup.flags {
+		if f.mask&(bitMask(1)<<index) != 0 {
+			if comma {
+				s = s + ", "
+			}
+			s = s + flag.name
+			comma = true
+		}
+	}
+	s = s + "]"
+	return s
+}
+
+// NewNodeStateMachine creates a new node state machine.
+// If db is not nil then the node states, fields and active timeouts are persisted.
+// Persistence can be enabled or disabled for each state flag and field.
+func NewNodeStateMachine(db ethdb.KeyValueStore, dbKey []byte, clock mclock.Clock, setup *Setup) *NodeStateMachine {
+	if setup.flags == nil {
+		panic("No state flags defined")
+	}
+	if len(setup.flags) > 8*int(unsafe.Sizeof(bitMask(0))) {
+		panic("Too many node state flags")
+	}
+	ns := &NodeStateMachine{
+		db:           db,
+		dbMappingKey: []byte(string(dbKey) + "mapping:"),
+		dbNodeKey:    []byte(string(dbKey) + "node:"),
+		clock:        clock,
+		setup:        setup,
+		nodes:        make(map[enode.ID]*nodeInfo),
+		stateNameMap: make(map[string]int),
+		fields:       make([]*fieldInfo, len(setup.fields)),
+		fieldNameMap: make(map[string]int),
+	}
+	for index, flag := range setup.flags {
+		if _, ok := ns.stateNameMap[flag.name]; ok {
+			panic("Node state flag name collision")
+		}
+		ns.stateNameMap[flag.name] = index
+		if flag.persistent {
+			ns.saveFlags |= bitMask(1) << index
+		}
+	}
+	for index, field := range setup.fields {
+		if _, ok := ns.fieldNameMap[field.name]; ok {
+			panic("Node field name collision")
+		}
+		ns.fields[index] = &fieldInfo{fieldDefinition: field}
+		ns.fieldNameMap[field.name] = index
+	}
+	return ns
+}
+
+// stateMask checks whether the set of flags belongs to the same setup and returns its internal bit mask
+func (ns *NodeStateMachine) stateMask(flags Flags) bitMask {
+	if flags.setup != ns.setup && flags.mask != 0 {
```

Comment:
This condition allows using `nodestate.Flags{}` as an empty set of flags anywhere. It makes sense because in other cases the reference ensures the correct interpretation of the flag bits but an empty set is pretty universal so there is no need to create it from the specific `nodestate.Setup`.

Reply:
make sense